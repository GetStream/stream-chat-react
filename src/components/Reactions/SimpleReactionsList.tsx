import React, { PropsWithChildren, Suspense, useState } from 'react';
import clsx from 'clsx';

import { useChatContext } from '../../context/ChatContext';
import { useEmojiContext } from '../../context/EmojiContext';
import { useMessageContext } from '../../context/MessageContext';

import { useEnterLeaveHandlers } from '../Tooltip/hooks';
import { useProcessReactions } from './hooks/useProcessReactions';

import { PopperTooltip } from '../Tooltip';

import type { NimbleEmojiProps } from 'emoji-mart';
import type { ReactionResponse } from 'stream-chat';

import type { DefaultStreamChatGenerics } from '../../types/types';
import type { ReactionEmoji } from '../Channel/emojiData';

type WithTooltipProps = {
  onMouseEnter: React.MouseEventHandler;
  onMouseLeave: React.MouseEventHandler;
  title: React.ReactNode;
};

// todo: merge with ReactionsList/ButtonWithTooltip
// avoiding breaking change of replacing <span> with <button>
const WithTooltip = ({
  children,
  onMouseEnter,
  onMouseLeave,
  title,
}: PropsWithChildren<WithTooltipProps>) => {
  const [referenceElement, setReferenceElement] = useState<HTMLSpanElement | null>(null);
  const { handleEnter, handleLeave, tooltipVisible } = useEnterLeaveHandlers({
    onMouseEnter,
    onMouseLeave,
  });

  const { themeVersion } = useChatContext('WithTooltip');

  return (
    <>
      {themeVersion === '2' && (
        <PopperTooltip referenceElement={referenceElement} visible={tooltipVisible}>
          {title}
        </PopperTooltip>
      )}
      <span onMouseEnter={handleEnter} onMouseLeave={handleLeave} ref={setReferenceElement}>
        {children}
      </span>
    </>
  );
};

export type SimpleReactionsListProps<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
> = {
  /** Additional props to be passed to the [NimbleEmoji](https://github.com/missive/emoji-mart/blob/master/src/components/emoji/nimble-emoji.js) component from `emoji-mart` */
  additionalEmojiProps?: Partial<NimbleEmojiProps>;
  /** Function that adds/removes a reaction on a message (overrides the function stored in `MessageContext`) */
  handleReaction?: (reactionType: string, event: React.BaseSyntheticEvent) => Promise<void>;
  /** An array of the own reaction objects to distinguish own reactions visually */
  own_reactions?: ReactionResponse<StreamChatGenerics>[];
  /** An object that keeps track of the count of each type of reaction on a message */
  reaction_counts?: { [key: string]: number };
  /** A list of the currently supported reactions on a message */
  reactionOptions?: ReactionEmoji[];
  /** An array of the reaction objects to display in the list */
  reactions?: ReactionResponse<StreamChatGenerics>[];
};

const UnMemoizedSimpleReactionsList = <
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
>(
  props: SimpleReactionsListProps<StreamChatGenerics>,
) => {
  const { handleReaction: propHandleReaction, ...rest } = props;

  const { Emoji, emojiConfig } = useEmojiContext('SimpleReactionsList');
  const { handleReaction: contextHandleReaction } = useMessageContext<StreamChatGenerics>(
    'SimpleReactionsList',
  );

  const {
    additionalEmojiProps,
    emojiData,
    getEmojiByReactionType,
    iHaveReactedWithReaction,
    latestReactions,
    latestReactionTypes,
    supportedReactionsArePresent,
    totalReactionCount,
  } = useProcessReactions({ emojiConfig, ...rest });

  const [tooltipReactionType, setTooltipReactionType] = useState<string | undefined>(undefined);
  const { themeVersion } = useChatContext('SimpleReactionsList');

  const handleReaction = propHandleReaction || contextHandleReaction;

  if (!latestReactions.length) return null;

  if (!supportedReactionsArePresent) return null;

  const getUsersPerReactionType = (type: string | undefined) =>
    latestReactions
      .map((reaction) => {
        if (type && reaction.type === type) {
          return reaction.user?.name || reaction.user?.id;
        }
        return null;
      })
      .filter(Boolean);

  return (
    <div className='str-chat__message-reactions-container'>
      <ul
        className='str-chat__simple-reactions-list str-chat__message-reactions'
        data-testid='simple-reaction-list'
        onMouseLeave={() => setTooltipReactionType(undefined)}
      >
        {latestReactionTypes.map((reactionType, i) => {
          const emojiObject = getEmojiByReactionType(reactionType);
          const isOwnReaction = iHaveReactedWithReaction(reactionType);
          const tooltipVisible = emojiObject && tooltipReactionType === emojiObject?.id;
          const tooltipContent = getUsersPerReactionType(tooltipReactionType)?.join(', ');

          return emojiObject ? (
            <li
              className={clsx('str-chat__simple-reactions-list-item', {
                'str-chat__message-reaction-own': isOwnReaction,
              })}
              key={`${emojiObject.id}-${i}`}
              onClick={(event) => handleReaction(reactionType, event)}
              onKeyUp={(event) => handleReaction(reactionType, event)}
            >
              <WithTooltip
                onMouseEnter={() => setTooltipReactionType(reactionType)}
                onMouseLeave={() => setTooltipReactionType(undefined)}
                title={tooltipContent}
              >
                {
                  <Suspense fallback={null}>
                    <Emoji
                      data={emojiData}
                      emoji={emojiObject}
                      size={13}
                      {...additionalEmojiProps}
                    />
                  </Suspense>
                }
                &nbsp;
                {tooltipVisible && themeVersion === '1' && (
                  <div className='str-chat__simple-reactions-list-tooltip'>
                    <div className='arrow' />
                    {tooltipContent}
                  </div>
                )}
              </WithTooltip>
            </li>
          ) : null;
        })}
        {
          <li className='str-chat__simple-reactions-list-item--last-number'>
            {totalReactionCount}
          </li>
        }
      </ul>
    </div>
  );
};

export const SimpleReactionsList = React.memo(
  UnMemoizedSimpleReactionsList,
) as typeof UnMemoizedSimpleReactionsList;
