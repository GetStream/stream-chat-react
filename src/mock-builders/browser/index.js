export * from './AnalyserNode';
export * from './AudioContext';
export * from './EventEmitter';
export * from './MediaRecorder';
export * from './ResizeObserver';
