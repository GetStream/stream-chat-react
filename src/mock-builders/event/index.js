export { default as dispatchChannelDeletedEvent } from './channelDeleted';
export { default as dispatchChannelHiddenEvent } from './channelHidden';
export { default as dispatchChannelTruncatedEvent } from './channelTruncated';
export { default as dispatchChannelUpdatedEvent } from './channelUpdated';
export { default as dispatchChannelVisibleEvent } from './channelVisible';
export { default as dispatchConnectionChangedEvent } from './connectionChanged';
export { default as dispatchConnectionRecoveredEvent } from './connectionRecovered';
export { default as dispatchMessageDeletedEvent } from './messageDeleted';
export { default as dispatchMessageNewEvent } from './messageNew';
export { default as dispatchMessageUpdatedEvent } from './messageUpdated';
export { default as dispatchNotificationAddedToChannelEvent } from './notificationAddedToChannel';
export { default as dispatchNotificationMessageNewEvent } from './notificationMessageNew';
export { default as dispatchNotificationMutesUpdated } from './notificationMutesUpdated';
export { default as dispatchNotificationRemovedFromChannel } from './notificationRemovedFromChannel';
export { default as dispatchUserUpdatedEvent } from './userUpdated';
