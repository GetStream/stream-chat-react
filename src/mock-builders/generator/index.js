export * from './attachment';
export * from './channel';
export * from './member';
export * from './message';
export * from './reaction';
export * from './user';
export * from './upload';
