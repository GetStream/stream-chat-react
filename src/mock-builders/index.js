/* eslint-disable no-param-reassign */
/* eslint-disable no-underscore-dangle */
import { StreamChat } from 'stream-chat';
import { nanoid } from 'nanoid';

const apiKey = 'API_KEY';
const token = 'dummy_token';

const connectUser = (client, user) =>
  new Promise((resolve) => {
    client.connectionId = 'dumm_connection_id';
    client.user = user;
    client.user.mutes = [];
    client._user = { ...user };
    client.userID = user.id;
    client.userToken = token;
    client.wsPromise = Promise.resolve(true);
    resolve();
  });

function mockClient(client) {
  jest.spyOn(client, '_setToken').mockImplementation();
  jest.spyOn(client, '_setupConnection').mockImplementation();
  jest.spyOn(client, '_setupConnection').mockImplementation();
  jest.spyOn(client, 'getAppSettings').mockImplementation();
  client.tokenManager = {
    getToken: jest.fn(() => token),
    tokenReady: jest.fn(() => true),
  };
  client.connectUser = connectUser.bind(null, client);
  return client;
}

export const getTestClient = () => mockClient(new StreamChat(apiKey));

export const getTestClientWithUser = async (user = { id: nanoid() }) => {
  const client = mockClient(new StreamChat(apiKey));
  await connectUser(client, user);
  return client;
};

export const getRandomInt = (min, max) => {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min)) + min; // The maximum is exclusive and the minimum is inclusive
};

export * from './api';
export * from './event';
export * from './generator';
export * from './translator';
export * from './utils';
