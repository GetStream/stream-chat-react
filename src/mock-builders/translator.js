export const mockTranslatorFunction = (value, params = {}) =>
  Object.keys(params).reduce((acc, key) => {
    const regexp = new RegExp(`\\{\\{\\s${key}\\s\\}\\}`, 'g');
    return acc.replace(regexp, params[key]);
  }, value);
