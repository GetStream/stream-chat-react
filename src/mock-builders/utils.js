import { generateChannel, generateMember, generateMessage, generateUser } from './generator';
import { getOrCreateChannelApi, getTestClientWithUser, useMockedApis } from './index';

export async function createClientWithChannel(
  { channelData, empty, existingClient, existingUsers, memberCount, messageCount } = {
    channelData: { image: 'image-xxx', name: 'channel-xxx' },
    empty: false,
    existingClient: null,
    existingUsers: null,
    memberCount: 2,
    messageCount: 20,
  },
) {
  const users = existingUsers || Array.from({ length: memberCount }, generateUser);
  const members = users.map((user) => generateMember({ user }));
  const mockedChannel = generateChannel({
    data: channelData,
    members,
    messages: empty
      ? []
      : ' '
          .repeat(messageCount)
          .split(' ')
          .map((v, i) => generateMessage({ user: users[i % memberCount] })),
  });

  const client = existingClient || (await getTestClientWithUser({ id: users[0].id }));
  useMockedApis(client, [getOrCreateChannelApi(mockedChannel)]); // eslint-disable-line react-hooks/rules-of-hooks
  const channel = client.channel('messaging', mockedChannel.id);
  await channel.watch();

  return { channel, client, users };
}

export const initClientWithChannels = async ({ channelsData, customUser } = {}) => {
  const user = customUser || generateUser();
  const client = await getTestClientWithUser(user);
  const defaultGenerateChannelOptions = {
    members: [generateMember({ user })],
  };
  const channels = await Promise.all(
    (channelsData ?? [defaultGenerateChannelOptions]).map(async (channelData) => {
      const mockedChannelData = generateChannel({
        ...defaultGenerateChannelOptions,
        ...channelData,
      });

      // eslint-disable-next-line react-hooks/rules-of-hooks
      useMockedApis(client, [getOrCreateChannelApi(mockedChannelData)]);
      const channel = client.channel(mockedChannelData.channel.type, mockedChannelData.channel.id);
      await channel.watch();
      jest.spyOn(channel, 'getConfig').mockImplementation(() => mockedChannelData.channel.config);
      return channel;
    }),
  );

  return { channels, client };
};
