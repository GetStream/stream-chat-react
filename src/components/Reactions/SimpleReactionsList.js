/* eslint-disable */
import React from 'react';
import PropTypes from 'prop-types';
import { NimbleEmoji } from 'emoji-mart';

import { defaultMinimalEmojis, emojiSetDef, emojiData } from '../../utils';

class SimpleReactionsList extends React.PureComponent {
  static propTypes = {
    reactions: PropTypes.array,
    /** Object/map of reaction id/type (e.g. 'like' | 'love' | 'haha' | 'wow' | 'sad' | 'angry') vs count */
    reaction_counts: PropTypes.object,
    showTooltip: PropTypes.bool,
    /** Provide a list of reaction options [{id: 'angry', emoji: 'angry'}] */
    reactionOptions: PropTypes.array,
    /**
     * Handler to set/unset reaction on message.
     *
     * @param type e.g. 'like' | 'love' | 'haha' | 'wow' | 'sad' | 'angry'
     * */
    handleReaction: PropTypes.func,
  };

  static defaultProps = {
    showTooltip: true,
    reactionOptions: defaultMinimalEmojis,
    emojiSetDef,
  };

  state = {
    showTooltip: false,
    users: [],
  };

  showTooltip = () => {
    this.setState({
      showTooltip: true,
    });
  };

  hideTooltip = () => {
    this.setState({
      showTooltip: false,
    });
  };

  handleReaction = (e, type) => {
    if (e !== undefined && e.preventDefault) {
      e.preventDefault();
    }
    this.props.handleReaction(type);
    this.setUsernames(type);
  };

  getReactionCount = () => {
    const reaction_counts = this.props.reaction_counts;
    let count = null;
    if (
      reaction_counts !== null &&
      reaction_counts !== undefined &&
      Object.keys(reaction_counts).length > 0
    ) {
      count = 0;
      Object.keys(reaction_counts).map(
        (key) => (count += reaction_counts[key]),
      );
    }
    return count;
  };

  renderUsers = (users) =>
    users.map((user, i) => {
      let text = user;
      if (i + 1 < users.length) {
        text += ', ';
      }
      return (
        <span className="latest-user-username" key={`key-${i}-${user}`}>
          {text}
        </span>
      );
    });

  getReactionsByType = (reactions) => {
    const reactionsByType = {};
    reactions.map((item) => {
      if (reactionsByType[item.type] === undefined) {
        return (reactionsByType[item.type] = [item]);
      } else {
        return (reactionsByType[item.type] = [
          ...reactionsByType[item.type],
          item,
        ]);
      }
    });
    return reactionsByType;
  };

  renderReactions = (reactions) => {
    const reactionsByType = this.getReactionsByType(reactions);
    const reactionsEmojis = this.props.reactionOptions.reduce(
      (acc, cur) => ({ ...acc, [cur.id]: cur }),
      {},
    );
    return Object.keys(reactionsByType).map((type, i) => (
      <li
        className="str-chat__simple-reactions-list-item"
        key={`${reactionsByType[type][0].id}-${i}`}
        onClick={(e) => this.handleReaction(e, type)}
      >
        <span onMouseEnter={() => this.setUsernames(type)}>
          <NimbleEmoji
            emoji={reactionsEmojis[type]}
            {...emojiSetDef}
            size={13}
            data={emojiData}
          />
          &nbsp;
        </span>
      </li>
    ));
  };

  getUsernames = (reactions) =>
    reactions.map((item) =>
      item.user !== null ? item.user.name || item.user.id : 'null',
    );

  setUsernames = (type) => {
    const reactionsByType = this.getReactionsByType(this.props.reactions);

    const reactions = reactionsByType[type];
    const users = this.getUsernames(reactions);

    this.setState(
      {
        users,
      },
      () => this.showTooltip(),
    );
  };

  renderUsernames = (users) => users.join(', ');

  render() {
    const { reactions } = this.props;
    if (!reactions || reactions.length === 0) {
      return null;
    }
    return (
      <ul
        data-testid="simple-reaction-list"
        className="str-chat__simple-reactions-list"
        onMouseLeave={this.hideTooltip}
      >
        {this.state.showTooltip && (
          <div
            className="str-chat__simple-reactions-list-tooltip"
            ref={this.reactionSelectorTooltip}
          >
            <div className="arrow" />
            {this.renderUsernames(this.state.users)}
          </div>
        )}
        {this.renderReactions(reactions)}
        {reactions.length !== 0 && (
          <li className="str-chat__simple-reactions-list-item--last-number">
            {this.getReactionCount()}
          </li>
        )}
      </ul>
    );
  }
}

export default SimpleReactionsList;
