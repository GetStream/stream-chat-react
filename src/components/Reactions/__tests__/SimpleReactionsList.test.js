import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { generateReaction } from 'mock-builders';
import { NimbleEmoji as EmojiComponentMock } from 'emoji-mart';
import SimpleReactionsList from '../SimpleReactionsList';

jest.mock('emoji-mart', () => ({
  NimbleEmoji: jest.fn(({ emoji }) => (
    <div data-testid={`emoji-${emoji.id}`} />
  )),
}));

const handleReactionMock = jest.fn();
const loveEmojiTestId = 'emoji-love';

const renderComponent = ({ reaction_counts = {}, ...props }) => {
  const reactions = Object.entries(reaction_counts)
    .map(([type, count]) =>
      Array(count)
        .fill()
        .map(() => generateReaction({ type })),
    )
    .flat();

  return {
    ...render(
      <SimpleReactionsList
        reaction_counts={reaction_counts}
        reactions={reactions}
        handleReaction={handleReactionMock}
        {...props}
      />,
    ),
    reactions,
  };
};

const expectEmojiToHaveBeenRendered = (id) => {
  expect(EmojiComponentMock).toHaveBeenCalledWith(
    expect.objectContaining({
      emoji: expect.objectContaining({ id }),
    }),
    {},
  );
};

describe('SimpleReactionsList', () => {
  afterEach(jest.clearAllMocks);

  it('should render the total reaction count', () => {
    const { getByText } = renderComponent({
      reaction_counts: {
        love: 5,
        angry: 2,
      },
    });
    const count = getByText('7');
    expect(count).toBeInTheDocument();
    expect(count).toHaveClass(
      'str-chat__simple-reactions-list-item--last-number',
    );
  });

  it('should render an emoji for each type of reaction', () => {
    const reaction_counts = {
      love: 5,
      angry: 2,
    };
    renderComponent({ reaction_counts });

    expect(EmojiComponentMock).toHaveBeenCalledTimes(
      Object.keys(reaction_counts).length,
    );

    Object.keys(reaction_counts).forEach(expectEmojiToHaveBeenRendered);
  });

  it('should handle custom reaction options', () => {
    const reaction_counts = {
      banana: 1,
      cowboy: 2,
    };

    renderComponent({
      reaction_counts,
      reactionOptions: [
        { id: 'banana', emoji: '🍌' },
        { id: 'cowboy', emoji: '🤠' },
      ],
    });

    expect(EmojiComponentMock).toHaveBeenCalledTimes(
      Object.keys(reaction_counts).length,
    );

    Object.keys(reaction_counts).forEach(expectEmojiToHaveBeenRendered);
  });

  it('should call handleReaction callback if a reaction emoji is clicked', () => {
    const reaction_counts = {
      love: 1,
    };

    const { getByTestId } = renderComponent({ reaction_counts });

    fireEvent.click(getByTestId(loveEmojiTestId));

    expect(handleReactionMock).toHaveBeenCalledWith('love');
  });

  it('should render a tooltip with all users that reacted a certain way if the emoji is hovered', () => {
    const reaction_counts = {
      love: 3,
    };

    const { reactions, getByTestId, queryByText } = renderComponent({
      reaction_counts,
    });

    fireEvent.mouseEnter(getByTestId(loveEmojiTestId));

    reactions.forEach(({ user }) => {
      expect(queryByText(user.id, { exact: false })).toBeInTheDocument();
    });

    fireEvent.mouseLeave(getByTestId(loveEmojiTestId));

    reactions.forEach(({ user }) => {
      expect(queryByText(user.id, { exact: false })).not.toBeInTheDocument();
    });
  });
});
