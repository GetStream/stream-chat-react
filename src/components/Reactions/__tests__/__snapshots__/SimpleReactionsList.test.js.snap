// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleReactionsList v1 should handle custom reaction options 1`] = `<div />`;

exports[`SimpleReactionsList v1 should render an emoji for each type of reaction 1`] = `
<div>
  <div
    class="str-chat__message-reactions-container"
  >
    <ul
      class="str-chat__simple-reactions-list str-chat__message-reactions"
      data-testid="simple-reaction-list"
    >
      <li
        class="str-chat__simple-reactions-list-item"
      >
        <span>
          😂
           
        </span>
      </li>
      <li
        class="str-chat__simple-reactions-list-item"
      >
        <span>
          ❤️
           
        </span>
      </li>
      <li
        class="str-chat__simple-reactions-list-item--last-number"
      >
        7
      </li>
    </ul>
  </div>
</div>
`;

exports[`SimpleReactionsList v2 should handle custom reaction options 1`] = `<div />`;

exports[`SimpleReactionsList v2 should render an emoji for each type of reaction 1`] = `
<div>
  <div
    class="str-chat__message-reactions-container"
  >
    <ul
      class="str-chat__simple-reactions-list str-chat__message-reactions"
      data-testid="simple-reaction-list"
    >
      <li
        class="str-chat__simple-reactions-list-item"
      >
        <span>
          😂
           
        </span>
      </li>
      <li
        class="str-chat__simple-reactions-list-item"
      >
        <span>
          ❤️
           
        </span>
      </li>
      <li
        class="str-chat__simple-reactions-list-item--last-number"
      >
        7
      </li>
    </ul>
  </div>
</div>
`;
